source/m3_api_libc.c

```diff
@@ -105,6 +104,121 @@ void internal_itoa(int n, char s[], int radix)
     }
 }
 
+static
+void internal_uitoa(uint32_t n, char s[], int radix, bool upper)
+{
+    static char const HEXDIGITS[0x10] = {
+            '0', '1', '2', '3', '4', '5', '6', '7',
+            '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
+    };
+
+    int i, j;
+    char c;
+
+    i = 0;
+    do {
+        char ch = HEXDIGITS[n % radix];
+        s[i++] = ch | ((!upper) << 5);
+    } while ((n /= radix) > 0);
+
+    s[i] = '\0';
+
+    // reverse
+    for (i = 0, j = strlen(s)-1; i<j; i++, j--) {
```

Comment:
Another simplification:
```suggestion
    for (i = 0, j = i - 1; i < j; i++, j--) {
```

Reply:
This doesn't work for me. Something got reversed. The test result becomes `Message 1: test 321- 321 dcba DCBA
` instead of `Message 1: test -123 123 abcd ABCD`. I think you need to reverse back again?

Reply:
Ah, sorry it should be:
```c
for (j = i - 1, i = 0; i < j; i++, j--) {
```

Reply:
Yep all good, thanks!